File: api/src/main/java/com/delivery/common/dto/CommonResponse.java
package com.delivery.common.dto;

import com.delivery.common.status.ErrorStatusCode;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Getter
@NoArgsConstructor
public class CommonResponse<T> {
    public CommonResponse(CommonMeta meta) {
        this.meta = meta;
    }

    public CommonResponse(CommonMeta meta, T body) {
        this.meta = meta;
        this.body = body;
    }

    private CommonMeta meta;

    private T body;

    public static <T> CommonResponse<T> ok() {
        return new CommonResponse<>(CommonMeta.ok());
    }

    public static <T> CommonResponse<T> ok(T body) {
        return new CommonResponse<>(CommonMeta.ok(), body);
    }

    public static CommonResponse<Object> ok(CommonMeta meta) {
        return new CommonResponse<>(meta);
    }

    public static CommonResponse<Object> error(ErrorStatusCode errorStatusCode) {
        return new CommonResponse<>(CommonMeta.error(errorStatusCode));
    }

    public static CommonResponse<Object> error(ErrorStatusCode errorStatusCode, Throwable tx) {
        return new CommonResponse<>(CommonMeta.error(errorStatusCode, tx));
    }

    public static CommonResponse<Object> error(ErrorStatusCode errorStatusCode, String description) {
        return new CommonResponse<>(CommonMeta.error(errorStatusCode, description));
    }
}

File: common/src/main/kotlin/com/delivery/common/dto/CommonResponse.kt
package com.delivery.common.dto

import com.delivery.common.status.StatusCode
import jakarta.validation.Valid

data class CommonResponse<T>(
    var commonMeta: CommonMeta,

    @field:Valid
    var body: T? = null
) {
    companion object {
        @JvmStatic
        fun <T> ok(): CommonResponse<T> {
            return CommonResponse(CommonMeta.ok())
        }

        @JvmStatic
        fun <T> ok(
            body: T
        ): CommonResponse<T> {
            return CommonResponse(CommonMeta.ok(), body)
        }

        @JvmStatic
        fun ok(
            meta: CommonMeta
        ): CommonResponse<Any> {
            return CommonResponse(meta)
        }

        @JvmStatic
        fun error(
            errorStatusCode: StatusCode
        ): CommonResponse<Any> {
            return CommonResponse(CommonMeta.error(errorStatusCode))
        }

        @JvmStatic
        fun error(
            errorStatusCode: StatusCode,
            t: Throwable,
        ): CommonResponse<Any> {
            return CommonResponse(CommonMeta.error(errorStatusCode, t))
        }

        @JvmStatic
        fun error(
            errorStatusCode: StatusCode,
            description: String,
        ): CommonResponse<Any> {
            return CommonResponse(CommonMeta.error(errorStatusCode, description))
        }
    }
}
